guides/keras_nlp/transformer_pretraining.py

```diff
@@ -0,0 +1,481 @@
+Modeling* (MLM) ojective.
+
+Essentially, we will be playing a big game of "guess the missing word." For each input
+sample we will obscure 20% of our input data, and train our model to predict the parts we
+covered up.
+
+Note that we aren't following any particular transformer architecture here, but you can
+think of our approach as a simplified
+[BERT](https://en.wikipedia.org/wiki/BERT_(language_model)) model.
+"""
+
+"""
+### Preprocess data for the MLM task
+
+Our text preprocessing for the MLM task will occur in two stages.
+ - First, we tokenize input text into integer sequences of token ids.
+ - Second, we mask certain positions in our input to predict on.
+
+To tokenize, we can use a `keras_nlp.tokenizer.Tokenizer`--the KerasNLP building block
+for transforming text into sequences of integer token ids.
+
+In particular, we will use `keras_nlp.tokenizers.WordPieceTokenizer` which does
+*sub-word* tokenization. Sub-word tokenization is popular when training models on large
+text corpora. It offers a good trade off--you don't need a massive vocabulary to encode
+every word in your corpus, but you can still encode common words in their entirety.
+
+The second thing we need to do is mask our input for the MLM task. To do this, we can use
+`keras_nlp.layers.MLMMaskGenerator`, which will randomly select a set of tokens in each
+input and mask them out.
+
+The tokenizer and the masking layer can both be used inside a call to
+[tf.data.Dataset.map](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#map).
+This will allow us to efficiently pre-compute each batch on the CPU, while our GPU or TPU
+works on training with the batch that came before. And because our masking layer will
+choose new words to mask each time, each epoch over our dataset will give us a totally
+new set of labels to train on.
+"""
+
+# Setting sequence_length will trim or pad the token outputs to shape
+# (batch_size, SEQ_LENGTH).
+tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary="bert_vocab_uncased.txt", sequence_length=SEQ_LENGTH,
+)
+# Setting mask_selection_length will trim or pad the mask outputs to shape
+# (batch_size, PREDICTIONS_PER_SEQ).
+masker = keras_nlp.layers.MLMMaskGenerator(
+    vocabulary_size=tokenizer.vocabulary_size(),
+    mask_selection_rate=MASK_RATE,
+    mask_selection_length=PREDICTIONS_PER_SEQ,
+    mask_token_id=tokenizer.token_to_id("[MASK]"),
+)
+
+
+def preprocess(inputs):
+    inputs = tokenizer(inputs)
+    outputs = masker(inputs)
+    # Split the masking layer outputs into a (features, labels, and weights)
+    # tuple that we can use with keras.Model.fit().
+    features = {
+        "tokens": outputs["tokens"],
+        "mask_positions": outputs["mask_positions"],
+    }
+    labels = outputs["mask_ids"]
+    weights = outputs["mask_weights"]
+    return features, labels, weights
+
+
+# We use prefetch() to pre-compute preprocessed batches on the fly on our CPU.
+pretrain_ds = wiki_train_ds.map(
```

Comment:
How large is it? In some cases we may want to recommend `cache()` so that we only do the preprocessing work once.

Reply:
Maybe 500mb?

But I don't think we can cache, because we want the masking to be a random augmentation not static. We could cache after tokenization, but I wouldn't want the complexity of breaking up the preprocess function. I can check if that's a speedup.

Reply:
Got it. Makes sense not to cache then.